packages/salesforcedx-vscode-apex/src/languageUtils/languageClientUtils.ts

```diff
@@ -104,3 +119,135 @@ export const getExceptionBreakpointInfo = async (): Promise<{}> => {
   }
   return Promise.resolve(response);
 };
+
+export const restartLanguageServerAndClient = async (extensionContext: vscode.ExtensionContext) => {
+  const removeIndexFiles = await vscode.window.showInformationMessage(
+    nls.localize('apex_language_server_restart_dialog_prompt'),
+    nls.localize('apex_language_server_restart_dialog_clean_and_restart'),
+    nls.localize('apex_language_server_restart_dialog_restart_only')
+  );
+  //get the client and status bar item
+  const alc = LanguageClientUtils.getInstance().getClientInstance();
+  const statusBarInstance = LanguageClientUtils.getInstance().getStatusBarInstance() ?? new ApexLSPStatusBarItem();
+  LanguageClientUtils.getInstance().setStatusBarInstance(statusBarInstance);
+
+  if (alc !== undefined) {
+    statusBarInstance.restarting(); // update the status to user that we're restarting
+    // Stop the client and server with the safe stop() routine from the client.
+    // Server is stopped automatically as part of that procedure
+    try {
+      //if we try and restart during another restart or with a startFailed status, we'll throw an exception
+      await alc.stop();
+    } catch (error) {
+      //if the client is in the state of startFailed, we need to kill the LSP jar and then move on.
+      //todo: do we want to kill LSP jar manually here?
+      vscode.window.showWarningMessage(
+        nls.localize('apex_language_server_restart_dialog_restart_only') + error.message
+      );
+    }
+    if (removeIndexFiles === nls.localize('apex_language_server_restart_dialog_clean_and_restart')) {
+      //delete the apex DB file if we need a clean slate
+      await removeApexDB();
+    }
+    //Need a half second timeout for the creation of a new language client so that we give the previous LS a moment to complete shutdown
+    setTimeout(createLanguageClient, 500, extensionContext, statusBarInstance);
+  }
+};
+
+const removeApexDB = async () => {
+  if (vscode.workspace.workspaceFolders) {
```

Comment:
another nit, we have hasRootWorkspace() in workspaceUtils

Reply:
I have added it in but I have to also use the direct check for the tsconfig to be okay with me referencing it a few lines later that way.